src/roles.js

```diff
@@ -270,31 +276,156 @@ export class Roles extends TypedEmitter {
-      if (authCoreId === this.#projectCreatorAuthCoreId) {
-        return CREATOR_ROLE
-      } else {
-        // When no role assignment exists, e.g. a newly added device which has
-        // not yet synced role records.
-        return NO_ROLE
+    if (result) return result
+
+    // The project creator will have the creator role
+    const authCoreId = await this.#coreOwnership.getCoreId(deviceId, 'auth')
+    if (authCoreId === this.#projectCreatorAuthCoreId) {
+      return CREATOR_MEMBERSHIP_RECORD
+    }
+
+    // When no role assignment exists, e.g. a newly added device which has
+    // not yet synced membership records.
+    return null
+  }
+
+  /**
+   * @param {ReadonlyDeep<MembershipRecord>} membershipRecord
+   * @returns {Promise<boolean>}
+   */
+  async #isRoleChainValid(membershipRecord) {
+    if (membershipRecord.roleId === LEFT_ROLE_ID) return true
+
+    /** @type {null | ReadonlyDeep<MembershipRecord>} */
+    let currentMembershipRecord = membershipRecord
+
+    while (currentMembershipRecord) {
+      const parentMembershipRecord = await this.#getParentMembershipRecord(
+        currentMembershipRecord
+      )
+      switch (parentMembershipRecord) {
+        case null:
+          break
+        case CREATOR_MEMBERSHIP_RECORD:
+          return true
+        default:
+          if (
+            !canAssign({
+              assigner: parentMembershipRecord,
+              assignee: currentMembershipRecord,
+            })
+          ) {
+            return false
+          }
+          break
       }
+
+      currentMembershipRecord = parentMembershipRecord
     }
 
-    const { roleId } = roleRecord
-    if (!isRoleId(roleId)) {
-      return BLOCKED_ROLE
+    return false
+  }
+
+  /**
+   * @param {ReadonlyDeep<MembershipRecord>} membershipRecord
+   * @returns {Promise<null | typeof CREATOR_MEMBERSHIP_RECORD | MembershipRecord>}
+   */
+  async #getParentMembershipRecord(membershipRecord) {
+    const {
+      coreDiscoveryKey: assignerCoreDiscoveryKey,
+      index: assignerIndexAtAssignmentTime,
+    } = parseVersionId(membershipRecord.versionId)
+
+    const isAssignedByProjectCreator =
+      assignerCoreDiscoveryKey.toString('hex') ===
+      this.#projectCreatorAuthCoreId
+    if (isAssignedByProjectCreator) return CREATOR_MEMBERSHIP_RECORD
+
+    const assignerCore = this.#coreManager.getCoreByDiscoveryKey(
+      assignerCoreDiscoveryKey
+    )
+    if (assignerCore?.namespace !== 'auth') return null
+
+    const assignerCoreId = assignerCore.key.toString('hex')
+    const assignerDeviceId = await this.#coreOwnership
+      .getOwner(assignerCoreId)
+      .catch(() => null)
+    if (!assignerDeviceId) return null
+
+    const latestMembershipRecord = await this.#getMembershipRecord(
+      assignerDeviceId
+    )
+
+    /** @type {null | typeof CREATOR_MEMBERSHIP_RECORD | MembershipRecord} */
+    let membershipRecordToCheck = latestMembershipRecord
+    while (membershipRecordToCheck) {
```

Comment:
This should have a max-depth check, to avoid a bug/hack that might create a circular reference so that this never ends.